core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java

```diff
@@ -132,69 +134,89 @@ public boolean execute(TestPlan testPlan, InfrastructureConfig infrastructureCon
         long startTime = System.currentTimeMillis();
 
         // Provision infrastructure
+        testPlan.setPhase(TestPlanPhase.INFRA_PHASE_STARTED);
+        testPlanUOW.persistTestPlan(testPlan);
         InfrastructureProvisionResult infrastructureProvisionResult = provisionInfrastructure(infrastructureConfig,
                 testPlan);
         //setup product performance dashboard
         if (infrastructureProvisionResult.isSuccess()) {
+            persistTestPlanPhase(testPlan, TestPlanPhase.INFRA_PHASE_SUCCEEDED);
             GrafanaDashboardHandler dashboardSetup = new GrafanaDashboardHandler(testPlan.getId());
             dashboardSetup.initDashboard();
         }
 
         // Create and set deployment.
-        DeploymentCreationResult deploymentCreationResult = createDeployment(testPlan,
-                infrastructureProvisionResult);
+        if (testPlan.getPhase().equals(TestPlanPhase.INFRA_PHASE_SUCCEEDED) &&
+                testPlan.getStatus().equals(TestPlanStatus.RUNNING)) {
+            persistTestPlanPhase(testPlan, TestPlanPhase.DEPLOY_PHASE_STARTED);
```

Comment:
If possible, I'd like to keep the #execute method as clean and short as possible. Technically, it should only have four code lines - infra create, deploy create, test run, infra destroy. We won't be able to go to that level, but let's see what we can do.

Reply:
Refactored as much as possible to accomplish that :)

Reply:
Nice.. That's a good design!